web_widget_float_formula/static/src/js/web_widget_float_formula.js

```diff
@@ -0,0 +1,103 @@
+/**
+*    Copyright GRAP
+*    Copyright 2016 LasLabs Inc.
+*    License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+**/
+
+odoo.define('web_widget_float_formula', function(require) {
+    "use strict";
+
+    var core = require('web.core');
+    var basic_fields = require('web.basic_fields');
+    var FieldFloat = basic_fields.FieldFloat;
+    FieldFloat.include({
+        start: function() {
+            this._super();
+            this.on('blurred', this, this._compute_result);
+            this.on('focused', this, this._display_formula);
+
+            this.setUpFocus();
+        },
+
+        setUpFocus: function() {
+            var self = this;
+            this.$el.on({
+                focus: function() { self.trigger('focused'); },
+                blur: function() { self.trigger('blurred'); }
+            });
+        },
+
+        commitChanges: function() {
+            this._compute_result();
+        },
+
+        _formula_text: '',
+
+        _clean_formula_text: function() {
+            this._formula_text = '';
+        },
+
+        _process_formula: function(formula) {
+            try{
+                formula = formula.toString();
+            } catch (ex) {
+                return false;
+            }
+            var clean_formula = formula.toString().replace(/^\s+|\s+$/g, '');
+            if (clean_formula[0] == '=') {
+                clean_formula = clean_formula.substring(1);
+                var myreg = new RegExp('[0-9]|\\s|\\.|,|\\(|\\)|\\+|\\-|\\*|\\/', 'g');
+                if (clean_formula.replace(myreg, '') === '') {
+                    return clean_formula;
+                }
+            }
+            return false;
+        },
+
+        _eval_formula: function(formula) {
+            // Import localization values used to eval formula
+            var translation_params = core._t.database.parameters;
+            var decimal_point = translation_params.decimal_point;
+            var thousands_sep = translation_params.thousands_sep;
+
+            var value;
+            formula = formula.replace(thousands_sep, '').replace(decimal_point, '.');
+            try {
+                value = eval(formula);
+            }
+            catch(e) {}
+
+            if (typeof value != 'undefined') {
+                return value;
+            }
+            return false;
+        },
+
+        _compute_result: function() {
+            this._clean_formula_text();
+
+            var input = this.$input.val();
```

Comment:
With float fields that are defined as computed fields I get this error while creating a record:

```
Error:
TypeError: this.$input is undefined

http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:79
Traceback:
_compute_result@http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:79:17
commitChanges@http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:31:13
commitChanges/defs<@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_renderer.js:67:20
_.collect@http://127.0.0.1:8069/web/static/lib/underscore/underscore.js:164:24
commitChanges@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_renderer.js:66:20
exec/<@http://127.0.0.1:8069/web/static/src/js/core/concurrency.js:216:31
then/</</<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3276:86
fire@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3119:38
add@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3165:49
then/</<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3275:77
each@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:383:49
then/<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3272:49
Deferred@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3332:25
then@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3271:48
exec@http://127.0.0.1:8069/web/static/src/js/core/concurrency.js:215:20
saveRecord@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_controller.js:167:16
saveRecord@http://127.0.0.1:8069/web/static/src/js/views/form/form_controller.js:191:16
OdooClass.extend/</prototype[name]</<@http://127.0.0.1:8069/web/static/src/js/core/class.js:90:35
_onSave@http://127.0.0.1:8069/web/static/src/js/views/form/form_controller.js:547:9
dispatch@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:4640:50
add/elemData.handle@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:4309:41
```

That error can be simply avoided by checking that `this.$input` is defined.

It seems to work fine if you write this block of code this way:

```
            if (typeof this.$input != 'undefined') {
                var input = this.$input.val();

                var formula = this._process_formula(input);
                if (formula !== false) {
                    var value = this._eval_formula(formula);
                    if (value !== false) {
                        this._formula_text = "=" + formula;

                        var decimal_point = core._t.database.parameters.decimal_point;

                        // _setValue
                        this._setValue(value.toString().replace(decimal_point, '.'))
                        this._render();
                    }
                }
            }
```

Could you apply this patch?